lib/active_shipping/carriers/ups.rb

```diff
@@ -275,53 +277,75 @@ def build_shipment_request(origin, destination, packages, options = {})
           end
 
           xml.Shipment do
-            # Required element.
             xml.Service do
-              xml.Code(options[:service_code] || '14')
-              xml.Description(options[:service_description] || 'Next Day Air Early AM')
+              xml.Code(options[:service_code] || '03')
             end
 
-            # Required element. The delivery destination.
-            build_location_node(xml, 'ShipTo', destination, {})
+            build_location_node(xml, 'ShipTo', destination, options)
+            build_location_node(xml, 'ShipFrom', origin, options)
             # Required element. The company whose account is responsible for the label(s).
-            build_location_node(xml, 'Shipper', options[:shipper] || origin, {})
-            # Required if pickup is different different from shipper's address.
-            build_location_node(xml, 'ShipFrom', options[:ship_from], {}) if options[:ship_from]
+            build_location_node(xml, 'Shipper', options[:shipper] || origin, options)
 
-            # Optional.
             if options[:saturday_delivery]
               xml.ShipmentServiceOptions do
                 xml.SaturdayDelivery
               end
             end
 
-            # Optional.
             if options[:origin_account]
               xml.RateInformation do
                 xml.NegotiatedRatesIndicator
               end
             end
 
-            # Optional.
-            if options[:shipment] && options[:shipment][:reference_number]
-              xml.ReferenceNumber do
-                xml.Code(options[:shipment][:reference_number][:code] || "")
-                xml.Value(options[:shipment][:reference_number][:value])
+            if options[:reference_numbers]
+              # UPS only allows shipment-level reference numbers if the package is not US -> US or PR -> PR
+              unless [['US','US'],['PR', 'PR']].include?([origin,destination].map(&:country_code))
```

Comment:
Also, unsure about this, but should we raise an exception (ArgumentError) if somebody specifies reference numbers, but UPS doesn't support it?